### ISSUE TITLE
Library cannot be loaded using dlopen()

### ISSUE THREAD

**Issue by robertwgh** [dlfcn-win32/dlfcn-win32#19]
I am trying to use dlfcn to load the OpenCL.dll library (which comes with Nvidia or AMD graphics driver).
OpenCL.dll is typically located at
`C:\Windows\System32\OpenCL.dll`, 32bit version
`C:\Windows\SysWOW64\OpenCL.dll`, 64bit version.

When I try to load the library, the `LoadLibraryEx()` gives an error code 126. I tried to figure out what happened. But haven't found out the reason yet. These two libraries are very common, and everyone has a Nvidia or AMD graphics card should already have these two dlls installed. 

Application runs well if I don't load the library manually and asks the application to find the library from the system path by itself.
Could you please help and see how to load this library correctly?
Thanks.

I attach the OpenCL.dll here in case you don't have the corresponding graphics card driver installed:
http://goo.gl/sy9F2C

**Comment by TimothyGu** [comment]
Can you give a reduced test case? What does `dlerror` output?

**Comment by robertwgh** [comment]
dlerror output is not complete for some reason, which makes it harder to
debug.
dlerror() shows "OpenCL.dll: T"

I don't know what that means.

**Comment by robertwgh** [comment]
The code I used is:

```C++
void * handle = = dlopen("C:\\Windows\\System32\\OpenCL.dll", RTLD_LAZY);
if( handle == NULL )
{
	printf("Error code %s", dlerror());
	return -1;
}
```

**Comment by TimothyGu** [comment]
What language is your Windows system running on? The message is supposed to be localized.

**Comment by TimothyGu** [comment]
Also, you shouldn't need to add the full path. Simply `dlopen("OpenCL.dll")` should work.

**Comment by robertwgh** [comment]
I think the problem might be because of the `LPSTR` type cast before the lpFileName:
```
hModule = LoadLibraryEx( (LPSTR) lpFileName, NULL, 
                                 LOAD_WITH_ALTERED_SEARCH_PATH );
```

While `LoadLibraryEx` is defined as follows:
```
#ifdef UNICODE
#define LoadLibraryEx  LoadLibraryExW
#else
#define LoadLibraryEx  LoadLibraryExA
#endif // !UNICODE
```

The language setting of my Windows is English. The non-unicode language is English, see 2.png.
So in my VS2012 project, I chose "Use Unicode charset",  see 1.png
then I got 126 error, and dlerror() reports "OpenCL.dll: T", see 3.png
Actually, when compiling, you will see the following warning:
>1>c:\github\openclz\testdll\testdll\dlfcn.c(170): warning C4133: 'function' : incompatible types - from 'char *' to 'LPWSTR'
1>c:\github\openclz\testdll\testdll\dlfcn.c(260): warning C4133: 'function' : incompatible types - from 'LPSTR' to 'LPCWSTR'


If I change the VS2012 settings to "Use multi-byte charset", then the dll can be opened correctly.

So, it seems that the dlfcn library should be revised to support both unicode and wide char. 


2.png
![2](https://cloud.githubusercontent.com/assets/1776084/7035541/7a33d3e0-dd3e-11e4-8757-2b9e79414bec.png)
1.png
![1](https://cloud.githubusercontent.com/assets/1776084/7035540/7a33af32-dd3e-11e4-8a82-1ebcf14e9898.png)



3.png
![3](https://cloud.githubusercontent.com/assets/1776084/7035539/7a332bc0-dd3e-11e4-8d36-29fd0c0f459f.png)

**Comment by robertwgh** [comment]
Change 

```
hModule = LoadLibraryEx( (LPSTR) lpFileName, NULL, 
                                 LOAD_WITH_ALTERED_SEARCH_PATH );
```
to
```
#ifdef UNICODE
		wchar_t wtext[1024];
		mbstowcs(wtext, lpFileName, strlen(lpFileName) + 1);
		LPWSTR lpwstr = wtext;
		hModule = LoadLibraryEx(lpwstr, NULL, 
			LOAD_WITH_ALTERED_SEARCH_PATH );
#else
		hModule = LoadLibraryEx( (LPCSTR) lpFileName, NULL, 
			LOAD_WITH_ALTERED_SEARCH_PATH );
#endif
```

By using this, I can open the library in Unicode mode. 
But at the same time, the following code gives error: `dlfcn.c(170): warning C4133: 'function' : incompatible types - from 'char *' to 'LPWSTR'`
```
pos += FormatMessage( FORMAT_MESSAGE_FROM_SYSTEM, NULL, dwMessageId,
                          MAKELANGID( LANG_NEUTRAL, SUBLANG_DEFAULT ),
                          error_buffer+pos, sizeof(error_buffer)-pos, NULL );
```

**Comment by robertwgh** [comment]
I can help fix these two place. Should I go ahead to start a pull request?

**Comment by TimothyGu** [comment]
Yes please!

**Comment by TimothyGu** [comment]
While at it you can change the default to unicode too.

**Comment by robertwgh** [comment]
Sure. I will work on it tonight. 
I will try to make as minimal changes as I can, but support both wchar and unicode.

**Comment by TimothyGu** [comment]
@robertwgh you know what you can just drop the support for multibyte. Nobody (should) use it anymore.

Also `wchar_t` is Unicode in Windows IIRC.

**Comment by robertwgh** [comment]
I think the current code works well for multibyte. We had better keep it
just in case someone need it for some legacy projects.
Plus, there are not much to modify. Probably just a few places.
Using #ifdef _UNICODE should be clean enough.

**Comment by TimothyGu** [comment]
Ok sounds good.

**Comment by TimothyGu** [comment]
Should be fixed in aa1401bf7cf45262f10c5318f99ce5be3ea767ae

**Comment by pali** [comment]
Solution in commit aa1401b is too long, complicated and adds lot of useless code. The main problem is that you should call `LoadLibraryExA` for Unicode builds (not `LoadLibraryEx` or `LoadLibraryExW`). Similarly with `FormatMessageA`. Fix is in pull request https://github.com/dlfcn-win32/dlfcn-win32/pull/50
